### PR TITLE
Add discrepancy analysis

### PR DESCRIPTION
**Pull request by tupui**
Add `SALib.analyze.discrepancy` which is the method described in https://arxiv.org/abs/2206.13470

Based on 2D sub projections of [Xi,Y], the discrepancy of each samples is calculated which gives a value for all Xi. This can be used as a sensitivity indice.

Opening this in draft to get early feedbacks on how to proceed next. It's mainly missing tests and doc, but the feature itself should already work.

cc @arnaldpuy

### CODE REVIEW

**Comment by ConnectedSystems** [comment]
Thanks @tupui !

cc @samanrazavi

I will take a look when I can over the weekend. Enjoy your Easter break!

**Comment by tupui** [comment]
Thanks you as well 😃

**Comment by ConnectedSystems** [comment]
Very clean implementation as always @tupui .

What would the preferred visualization be for this? Scatter plot with running mean/median?

**Comment by tupui** [comment]
Thanks!

> What would the preferred visualization be for this? Scatter plot with running mean/median?

I think there are 2 options. First is to use the default bar plot as Sobol' and second is to show a grid of canonical plots since this is how we get the indices in the background. Then we could show on each subplot the discrepancy value and do some highlighting of the highest/lowest values. And if we were to add the method we used in the article, then we could show a grid on the subplots and colour/mark the empty cells.

> Actually one other question: how would grouped factors be dealt with?

I am not sure you can do much since you are looking at the input vs output as opposed to just the output for Sobol'. Or maybe merge the values for a given group?? @arnaldpuy?

**Comment by tupui** [comment]
Hey @ConnectedSystems how do you want me to proceed here?

**Comment by ConnectedSystems** [comment]
To support analysis of grouped factors, I suggest simply taking the mean of $S$ for the group. This is what I decided to do for PAWN.

The docstring also needs to be expanded to give an overview of the method. Some hints on interpreting results or pitfalls in analysis would also be appreciated.

**Comment by tupui** [comment]
Sorry for the delay here. I think I addressed your comments.

File: src/SALib/analyze/discrepancy.py
Comment (ConnectedSystems) [inline src/SALib/analyze/discrepancy.py:27]: I prefer `np.ndarray` for detailing the argument type, but is specifying the full name the accepted norm? I'll defer to what you say here.

The descriptor holds redundant info as well. I would state what the array holds directly instead, e.g., "An array of model inputs" (the type is already detailed)

File: src/SALib/analyze/discrepancy.py
Comment (ConnectedSystems) [inline src/SALib/analyze/discrepancy.py:52]:
```suggestion
    uniformly covered as the quality of the measure depends on the value of
```

File: src/SALib/analyze/discrepancy.py
Comment (ConnectedSystems) [inline src/SALib/analyze/discrepancy.py:87]: Was the extra empty line necessary for deployed docs?

File: src/SALib/analyze/discrepancy.py
Comment (ConnectedSystems) [inline src/SALib/analyze/discrepancy.py:44]:
```suggestion
    as a sensitivity index.
```

Or a slightly more verbose option:


```
    as a measure of sensitivity.
```

**Review by ConnectedSystems — CHANGES_REQUESTED** [review]
Thanks @tupui 

Apologies, I cannot remember what we discussed earlier regarding plotting.
Was the default bar chart sufficient? I think I was intending to add a scatter plot with moving mean like in the pre-print you shared, but not sure if that is necessary or not.

If the bar chart is okay, I think we can merge this after my comments here are considered.

File: src/SALib/analyze/discrepancy.py
Comment (tupui) [reply src/SALib/analyze/discrepancy.py]: So we need the full name so that the cross links are working. This is because in `intersphinx_mapping` we use the key `numpy`. We could change this to `np` if you want (and update everywhere, I've seen that we use `numpy.array` sometimes and this should be `numpy.ndarray` at least.)

https://numpydoc.readthedocs.io/en/latest/format.html#parameters

**Review by tupui — COMMENTED** [review]
Thanks Takuya, I think I addressed your comments :)

**Comment by tupui** [comment]
Hope you don't mind, I just fixed here a few doc warning/issues

File: src/SALib/analyze/discrepancy.py
Comment (ConnectedSystems) [reply src/SALib/analyze/discrepancy.py]: Okay, I prefer `np` (personal preference) but I will leave that as a separate task for the future.

**Review by ConnectedSystems — APPROVED** [review]

**Comment by ConnectedSystems** [comment]
Thanks @tupui, happy for this to be merged now. I'll leave expanded support for visualization for later.

**Pull request merged**
